server/balancer.go

```diff
@@ -82,98 +94,168 @@ func (s *storageBalancer) GetResourceKind() ResourceKind {
 }
 
 func (s *storageBalancer) Schedule(cluster *clusterInfo) Operator {
-	region, source, target := scheduleStorage(cluster, s.opt, s.selector)
+	// Select a peer from the store with largest storage ratio.
+	region, oldPeer := scheduleRemovePeer(cluster, s.selector)
 	if region == nil {
 		return nil
 	}
 
-	diff := source.storageRatio() - target.storageRatio()
-	if diff < s.opt.GetMinBalanceDiffRatio() {
+	// We don't schedule region with abnormal number of replicas.
+	if len(region.GetPeers()) != s.rep.GetMaxReplicas() {
 		return nil
 	}
 
-	oldPeer := region.GetStorePeer(source.GetId())
-	newPeer, err := cluster.allocPeer(target.GetId())
-	if err != nil {
-		log.Errorf("failed to allocate peer: %v", err)
+	op := s.transferPeer(cluster, region, oldPeer)
+	if op == nil {
+		// We can't transfer peer from this store now, so we add it to the cache
+		// and skip it for a while.
+		s.cache.set(oldPeer.GetStoreId())
+	}
+	return op
+}
+
+func (s *storageBalancer) transferPeer(cluster *clusterInfo, region *regionInfo, oldPeer *metapb.Peer) Operator {
+	stores := cluster.getRegionStores(region)
+	source := cluster.getStore(oldPeer.GetStoreId())
+
+	// Allocate a new peer from the store with smallest storage ratio.
+	// We need to ensure the target store will not break the replication constraints.
+	excluded := newExcludedFilter(nil, region.GetStoreIds())
+	replication := newReplicationFilter(s.rep, stores, source)
+	newPeer := scheduleAddPeer(cluster, s.selector, excluded, replication)
+	if newPeer == nil {
+		return nil
+	}
+
+	target := cluster.getStore(newPeer.GetStoreId())
+	if source.storageRatio()-target.storageRatio() < s.opt.GetMinBalanceDiffRatio() {
 		return nil
 	}
 
 	return newTransferPeer(region, oldPeer, newPeer)
 }
 
-// replicaChecker ensures region has enough replicas.
+// replicaChecker ensures region has the best replicas.
 type replicaChecker struct {
-	cluster  *clusterInfo
-	opt      *scheduleOption
-	selector Selector
+	opt     *scheduleOption
+	rep     *Replication
+	cluster *clusterInfo
+	filters []Filter
 }
 
-func newReplicaChecker(cluster *clusterInfo, opt *scheduleOption) *replicaChecker {
+func newReplicaChecker(opt *scheduleOption, cluster *clusterInfo) *replicaChecker {
 	var filters []Filter
 	filters = append(filters, newStateFilter(opt))
 	filters = append(filters, newSnapshotCountFilter(opt))
 
 	return &replicaChecker{
-		cluster:  cluster,
-		opt:      opt,
-		selector: newBalanceSelector(storageKind, filters),
+		opt:     opt,
+		rep:     opt.GetReplication(),
+		cluster: cluster,
+		filters: filters,
 	}
 }
 
 func (r *replicaChecker) Check(region *regionInfo) Operator {
-	// If we have bad peer, we remove it first.
-	for _, peer := range r.collectBadPeers(region) {
-		return newRemovePeer(region, peer)
+	if op := r.checkDownPeer(region); op != nil {
+		return op
+	}
+	if op := r.checkOfflinePeer(region); op != nil {
+		return op
```

Comment:
Down means that the peer is not working for a while, we will remove the down peer first, and add a new peer later.
Offline means that we want to take a store offline, but the store and its peers are still working, so we can add a new peer otherwhere then remove the offline peer.